CredentialProvider.Microsoft/CredentialProviders/Vsts/MSAL/MsalTokenProvider.cs

```diff
@@ -62,46 +71,105 @@ public async Task<IMsalToken> AcquireTokenWithDeviceFlowAsync(Func<DeviceCodeRes
+        {
+            IPublicClientApplication publicClient = await GetPCAAsync(useLocalHost: false).ConfigureAwait(false);
+            var accounts = new List<IAccount>();
 
+            string loginHint = EnvUtil.GetMsalLoginHint();
+
+            accounts.AddRange(await publicClient.GetAccountsAsync());
+
+            if (this.brokerEnabled && accounts.Count == 0 && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                accounts.Add(PublicClientApplication.OperatingSystemAccount);
+            }
+
+            Guid? authorityTenant = null;
             try
             {
-                var msalBuilder = publicClient.AcquireTokenWithDeviceCode(new string[] { resource }, deviceCodeHandler);
-                var result = await msalBuilder.ExecuteAsync(linkedCancellationToken);
-                return new MsalToken(result);
+                authorityTenant = Guid.Parse((new Uri(this.authority)).AbsolutePath.Trim('/'));
+                this.Logger.Verbose($"Found tenant `{authorityTenant.Value}` authority URL: `{this.authority}`.");
+
             }
-            finally
+            catch (Exception e)
             {
-                var helper = await GetMsalCacheHelperAsync();
-                helper?.UnregisterCache(publicClient.UserTokenCache);
+                this.Logger.Verbose($"Could not determine tenant from authority URL `{this.authority}`: {e.Message}");
             }
-        }
 
-        public async Task<IMsalToken> AcquireTokenSilentlyAsync(CancellationToken cancellationToken)
-        {
-            var publicClient = await GetPCAAsync().ConfigureAwait(false);
-            var accounts = await publicClient.GetAccountsAsync();
+            var sortedAccounts = PrioritizeAccounts(accounts, authorityTenant, loginHint);
 
-            try
+            foreach ((IAccount account, string canonicalName, _) in sortedAccounts)
             {
-                foreach (var account in accounts)
-                {
-                    try
-                    {
-                        var silentBuilder = publicClient.AcquireTokenSilent(new string[] { resource }, account);
-                        var result = await silentBuilder.ExecuteAsync(cancellationToken);
-                        return new MsalToken(result);
-                    }
-                    catch (MsalUiRequiredException)
-                    { }
-                    catch (MsalServiceException)
-                    { }
-                }
+                this.Logger.Verbose($"Found in cache: {canonicalName}");
             }
-            finally
+
+            foreach ((IAccount account, string canonicalName, _) in sortedAccounts)
             {
-                var helper = await GetMsalCacheHelperAsync();
-                helper?.UnregisterCache(publicClient.UserTokenCache);
+                try
+                {
+                    this.Logger.Verbose($"Attempting to use identity `{canonicalName}`.");
```

Comment:
Does this actually reach out to AzDO to see if the identity has access?

Reply:
If you're referring to the MSAL acquire token then, no, it won't. Only when that's used. If there are multiple identities found that could reasonably connect (like corp + alt account, multiple MSAs, etc.) then we'd either need to try and use them, or have the user provide the logon hint in those cases. Could image there would be cases where both have "access", but really only one has the right set of permissions for the specific feed being used.